WordPress/Classes/Services/NotificationSyncService.swift

```diff
@@ -0,0 +1,304 @@
+import Foundation
+
+
+// MARK: - Notifications
+//
+let NotificationSyncServiceDidUpdateNotifications = "NotificationSyncServiceDidUpdateNotifications"
+
+
+// MARK: - NotificationSyncService
+//
+class NotificationSyncService
+{
+    /// Returns the Main Managed Context
+    ///
+    private var contextManager = ContextManager.sharedInstance()
+
+    /// Sync Service Remote
+    ///
+    private var remote: NotificationSyncServiceRemote!
+
+    /// Maximum number of Notes to Sync
+    ///
+    private let maximumNotes = 100
+
+
+    /// Designed Initializer
+    ///
+    init?() {
+        guard let dotcomAPI = dotcomAPI else {
+            return nil
+        }
+
+        remote = NotificationSyncServiceRemote(wordPressComRestApi: dotcomAPI)
+    }
+
+    /// Initializer: Useful for Unit Testing
+    ///
+    /// - Parameters:
+    ///     - manager: ContextManager Instance
+    ///     - wordPressComRestApi: The WordPressComRestApi that should be used.
+    ///
+    init?(manager: ContextManager, dotcomAPI: WordPressComRestApi) {
+        guard dotcomAPI.hasCredentials() else {
+            return nil
+        }
+
+        contextManager = manager
+        remote = NotificationSyncServiceRemote(wordPressComRestApi: dotcomAPI)
+    }
+
+
+
+    /// Syncs the latest *maximumNotes*:
+    ///
+    /// - Latest 100 hashes are retrieved (++efficiency++)
+    /// - Only those Notifications that were remotely changed (Updated / Inserted) will be retrieved
+    /// - Local collection will be updated. Old notes will be purged!
+    ///
+    func sync(completion: ((ErrorType?, Bool) -> Void)? = nil) {
+        assert(NSThread.isMainThread())
+
+        remote.loadHashes(withPageSize: maximumNotes) { error, remoteHashes in
+            guard let remoteHashes = remoteHashes else {
+                completion?(error, false)
+                return
+            }
+
+            self.deleteLocalMissingNotes(from: remoteHashes) {
```

Comment:
This is more of an enhancement suggestion than a blocker for this PR, but I've seen this pattern before (as recently as last week) and I think it'd be helpful to abstract it out. In essence, you have two sets of objects that you have to reconcile, but since replacing the local with the remote every time is a problem you have to calculate a series of operations to apply on the original set. This could be core data updates, or mapping changed attributes to DOM changes.

The problem I noticed in this particular code is that you have 5 nested callbacks and it wasn't obvious at first why the local operations weren't synchronous.

What I'd suggest is a more generic `diff` algorithm that:
1. Fetches the pairs `(id, hash)` for all local notes
2. Compares the local set to the remote set in one pass, returning separate sets of `deleted`, `updated`, and `inserted` notes.
3. Deletes the missing notes
4. Fires the request to load all the updated/inserted notes.

You can see we have a similar thing going on for [mergeBlogs](https://github.com/wordpress-mobile/WordPress-iOS/blob/c93e0211000bc6fdd1acb215875be16b955b467c/WordPress/Classes/Services/BlogService.m#L570). I'd love to just have a `Dictionary.diff` method.